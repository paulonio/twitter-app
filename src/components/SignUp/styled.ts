import { styled } from 'styled-components';

export const SignUpWrapper = styled.section`
  width: 100%;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const FormTitle = styled.h2`
  margin-top: 106px;
  margin-bottom: 35px;
  font-size: 30px;
  line-height: 40px;
  font-weight: 700;
`;

export const Form = styled.form`
  width: 40%;
  max-width: 750px;
  padding: 30px 40px 40px;
  display: flex;
  flex-direction: column;
`;

export const Input = styled.input`
  width: 100%;
  margin-top: 25px;
  padding: 23px 20px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
  color: rgba(0, 0, 0, 0.6);
  border: solid 1px rgba(0, 0, 0, 0.2);
  border-radius: 6px;
`;

export const ButtonWrapper = styled.div`
  margin-top: 35px;
`;

export const RecaptchaContainer = styled.div``;
