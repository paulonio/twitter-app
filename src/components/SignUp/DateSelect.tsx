import React, { FC } from 'react';

import { SELECT_DATE_DATA } from '@src/constants/constants';

import type { SelectProps } from './DatePickers';

import { Select, Option } from './styled';

const DateSelect: FC<SelectProps> = ({ register }) => {
  return (
    <Select {...register('date')}>
      {SELECT_DATE_DATA.map((date, idx) => (
        <Option key={date} value={date} disabled={idx === 0}>
          {date}
        </Option>
      ))}
    </Select>
  );
};

export default DateSelect;
