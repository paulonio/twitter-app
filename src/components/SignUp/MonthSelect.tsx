import React, { FC } from 'react';

import { SELECT_MONTH_DATA } from '@src/constants/constants';

import type { SelectProps } from './DatePickers';

import { Select, Option } from './styled';

const MonthSelect: FC<SelectProps> = ({ register }) => {
  return (
    <Select {...register('month')}>
      {SELECT_MONTH_DATA.map((month, idx) => (
        <Option key={month} value={month} disabled={idx === 0}>
          {month}
        </Option>
      ))}
    </Select>
  );
};

export default MonthSelect;
