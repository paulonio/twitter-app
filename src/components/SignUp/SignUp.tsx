import React, { ChangeEvent, FormEventHandler, useState } from 'react';
import {
  ConfirmationResult,
  RecaptchaVerifier,
  signInWithPhoneNumber,
  updatePassword,
  updateProfile,
} from 'firebase/auth';
import { useNavigate } from 'react-router-dom';
import { ButtonWrapper, Form, FormTitle, Input, RecaptchaContainer, SignUpWrapper } from './styled';
import Button from '../Button/Button';
import { auth } from '../../../firebase';
import { SecondaryButton } from '../SidebarUsers/styled';

declare const window: Window &
  typeof globalThis & {
    recaptchaVerifier: RecaptchaVerifier;
    confirmationResult: ConfirmationResult;
  };

export const generateRecaptcha = () => {
  window.recaptchaVerifier = new RecaptchaVerifier(
    'recaptcha-container',
    {
      size: 'invisible',
      callback: () => {},
    },
    auth
  );
};

const SignUp = () => {
  const navigate = useNavigate();
  const [name, setName] = useState<string>('');
  const [phone, setPhone] = useState<string>('');
  const [OTP, setOTP] = useState<string>('');
  const [password, setPassword] = useState<string>('');

  const [state, setState] = useState<'number' | 'verification' | 'password'>('number');

  const handleNavigateToWelcomePage = () => {
    navigate('/');
  };

  const onSubmit: FormEventHandler<HTMLFormElement> = async (e) => {
    e.preventDefault();
    generateRecaptcha();
    const appVerifier = window.recaptchaVerifier;

    await signInWithPhoneNumber(auth, phone, appVerifier)
      .then((confirmationResult: ConfirmationResult) => {
        window.confirmationResult = confirmationResult;
      })
      .catch(() => {});

    setState('verification');
  };

  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setName(value);
  };

  const handlePhoneChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setPhone(value);
  };

  const handlePasswordChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setPassword(value);
  };

  const verifyOTP = (e: ChangeEvent<HTMLInputElement>) => {
    const otp = e.target.value;
    setOTP(otp);

    if (otp.length === 6) {
      const confirmationResult = window.confirmationResult;
      confirmationResult
        .confirm(otp)
        .then((result) => {
          const user = result.user;
          updateProfile(user, { displayName: name }).catch(() => {});
        })
        .catch(() => {});
    }

    setState('password');
  };

  const handleAddPassword = () => {
    const user = auth.currentUser;
    if (user) {
      updatePassword(user, password);
      navigate('/feed');
    }
  };

  return (
    <SignUpWrapper>
      <RecaptchaContainer id="recaptcha-container" />
      <Form onSubmit={onSubmit}>
        <FormTitle>Create an account</FormTitle>
        {state === 'number' && (
          <>
            <Input type="text" placeholder="Name" value={name} onChange={handleNameChange} />
            <Input
              type="text"
              placeholder="Phone number"
              value={phone}
              onChange={handlePhoneChange}
            />
            <SecondaryButton onClick={handleNavigateToWelcomePage}>Use email</SecondaryButton>
          </>
        )}
        {state === 'verification' && (
          <Input type="text" placeholder="OTP" value={OTP} onChange={verifyOTP} />
        )}
        {state === 'password' && (
          <Input
            type="password"
            placeholder="Password"
            value={password}
            onChange={handlePasswordChange}
          />
        )}
        {state === 'password' ? (
          <ButtonWrapper onClick={handleAddPassword}>
            <Button buttonType="primary">Submit</Button>
          </ButtonWrapper>
        ) : (
          <ButtonWrapper>
            <Button buttonType="primary">Next</Button>
          </ButtonWrapper>
        )}
      </Form>
    </SignUpWrapper>
  );
};

export default SignUp;
