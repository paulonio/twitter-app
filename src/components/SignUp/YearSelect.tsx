import React, { FC } from 'react';

import { SELECT_YEAR_DATA } from '@src/constants/constants';

import type { SelectProps } from './DatePickers';

import { Select, Option } from './styled';

const YearSelect: FC<SelectProps> = ({ register }) => {
  return (
    <Select {...register('year')}>
      {SELECT_YEAR_DATA.map((year, idx) => (
        <Option key={year} value={year} disabled={idx === 0}>
          {year}
        </Option>
      ))}
    </Select>
  );
};

export default YearSelect;
