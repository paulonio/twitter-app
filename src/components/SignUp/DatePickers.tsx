import React, { FC } from 'react';
import { FieldErrors, UseFormRegister } from 'react-hook-form';

import MonthSelect from './MonthSelect';
import DateSelect from './DateSelect';
import YearSelect from './YearSelect';

import type { SignUpForm } from './SignUp';

import { SelectWrapper, HalfSelectWrapper, QuarterSelectWrapper, Error } from './styled';

export interface SelectProps {
  errors?: FieldErrors<SignUpForm>;
  register: UseFormRegister<SignUpForm>;
}

const DatePickers: FC<SelectProps> = ({ errors, register }) => {
  return (
    <SelectWrapper>
      <HalfSelectWrapper>
        <MonthSelect register={register} />
        <Error>{errors?.month?.message}</Error>
      </HalfSelectWrapper>
      <QuarterSelectWrapper>
        <DateSelect register={register} />
        <Error>{errors?.date?.message}</Error>
      </QuarterSelectWrapper>
      <QuarterSelectWrapper>
        <YearSelect register={register} />
        <Error>{errors?.year?.message}</Error>
      </QuarterSelectWrapper>
    </SelectWrapper>
  );
};

export default DatePickers;
