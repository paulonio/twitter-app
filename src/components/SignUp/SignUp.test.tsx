import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import SignUp from '@components/SignUp/SignUp';
import CustomProvider from '../EditUser/testUtils';

describe('Sign up', () => {
  it('should render sign up component', () => {
    const { container } = render(
      <CustomProvider>
        <SignUp />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should change sign up inputs', () => {
    render(
      <CustomProvider>
        <SignUp />
      </CustomProvider>
    );

    const nameInput: HTMLInputElement = screen.getByTestId('name');
    const emailInput: HTMLInputElement = screen.getByTestId('email');
    const passwordInput: HTMLInputElement = screen.getByTestId('password');
    fireEvent.change(nameInput, { target: { value: 'name' } });
    fireEvent.change(emailInput, { target: { value: 'test@test.com' } });
    fireEvent.change(passwordInput, { target: { value: 'test' } });
    expect(nameInput.value).toEqual('name');
    expect(emailInput.value).toEqual('test@test.com');
    expect(passwordInput.value).toEqual('test');
  });

  it('should set errors on incorrect values', async () => {
    render(
      <CustomProvider>
        <SignUp />
      </CustomProvider>
    );

    const nameInput: HTMLInputElement = screen.getByTestId('name');
    const emailInput: HTMLInputElement = screen.getByTestId('email');
    const passwordInput: HTMLInputElement = screen.getByTestId('password');
    const submit: HTMLButtonElement = screen.getByTestId('submit-button');
    fireEvent.change(nameInput, { target: { value: 'na' } });
    fireEvent.change(emailInput, { target: { value: 'test' } });
    fireEvent.change(passwordInput, { target: { value: 'test' } });
    fireEvent.click(submit);
    await waitFor(() => {
      expect(screen.getByText('name must be at least 3 characters')).toBeInTheDocument();
      expect(screen.getByText('email must be a valid email')).toBeInTheDocument();
      expect(screen.getByText('password must be at least 6 characters')).toBeInTheDocument();
    });
  });

  it('should change selects correctly', () => {
    render(
      <CustomProvider>
        <SignUp />
      </CustomProvider>
    );

    const select: HTMLSelectElement = screen.getByTestId('select');
    fireEvent.change(select, { target: { value: 'Apr' } });
    expect(select.value).toEqual('Apr');
  });
});
