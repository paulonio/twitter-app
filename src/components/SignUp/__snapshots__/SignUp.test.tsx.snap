// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up should render sign up component 1`] = `
<div>
  <section
    class="sc-hzJlrQ qlouH"
  >
    <img
      alt="Twitter logo"
      class="sc-cQdTPf epjNjn"
      src="/assets/icons/twitter-logo.svg"
    />
    <form
      class="sc-hlCNGy bdhUby"
    >
      <h2
        class="sc-hBNyCF dooLDi"
      >
        Create an account
      </h2>
      <input
        class="sc-hnHaRn eKyvOD"
        data-testid="name"
        name="name"
        placeholder="Name"
        type="text"
      />
      <p
        class="sc-SYvCl csQMFq"
      />
      <input
        class="sc-hnHaRn eKyvOD"
        data-testid="email"
        name="email"
        placeholder="Email"
        type="text"
      />
      <p
        class="sc-SYvCl csQMFq"
      />
      <button
        class="sc-erFrCC fpyDud"
      >
        Use email
      </button>
      <div
        class="sc-cEPTvB ilOSst"
      >
        <div
          class="sc-fPKvMM jRDJCo"
        >
          <select
            class="sc-goDJBy iWXaNu"
            data-testid="select"
            name="month"
          >
            <option
              class="sc-gqHWMn gwEFlI"
              disabled=""
              value="Month"
            >
              Month
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Jan"
            >
              Jan
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Feb"
            >
              Feb
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Mar"
            >
              Mar
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Apr"
            >
              Apr
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="May"
            >
              May
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Jun"
            >
              Jun
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Jul"
            >
              Jul
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Aug"
            >
              Aug
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Sep"
            >
              Sep
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Oct"
            >
              Oct
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Nov"
            >
              Nov
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="Dec"
            >
              Dec
            </option>
          </select>
          <p
            class="sc-SYvCl csQMFq"
          />
        </div>
        <div
          class="sc-jySWPz honjzI"
        >
          <select
            class="sc-goDJBy iWXaNu"
            name="date"
          >
            <option
              class="sc-gqHWMn gwEFlI"
              disabled=""
              value="Date"
            >
              Date
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1"
            >
              1
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2"
            >
              2
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="3"
            >
              3
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="4"
            >
              4
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="5"
            >
              5
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="6"
            >
              6
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="7"
            >
              7
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="8"
            >
              8
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="9"
            >
              9
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="10"
            >
              10
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="11"
            >
              11
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="12"
            >
              12
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="13"
            >
              13
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="14"
            >
              14
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="15"
            >
              15
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="16"
            >
              16
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="17"
            >
              17
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="18"
            >
              18
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="19"
            >
              19
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="20"
            >
              20
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="21"
            >
              21
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="22"
            >
              22
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="23"
            >
              23
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="24"
            >
              24
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="25"
            >
              25
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="26"
            >
              26
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="27"
            >
              27
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="28"
            >
              28
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="29"
            >
              29
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="30"
            >
              30
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="31"
            >
              31
            </option>
          </select>
          <p
            class="sc-SYvCl csQMFq"
          />
        </div>
        <div
          class="sc-jySWPz honjzI"
        >
          <select
            class="sc-goDJBy iWXaNu"
            name="year"
          >
            <option
              class="sc-gqHWMn gwEFlI"
              disabled=""
              value="Year"
            >
              Year
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1990"
            >
              1990
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1991"
            >
              1991
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1992"
            >
              1992
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1993"
            >
              1993
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1994"
            >
              1994
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1995"
            >
              1995
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1996"
            >
              1996
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1997"
            >
              1997
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1998"
            >
              1998
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="1999"
            >
              1999
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2000"
            >
              2000
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2001"
            >
              2001
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2002"
            >
              2002
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2003"
            >
              2003
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2004"
            >
              2004
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2005"
            >
              2005
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2006"
            >
              2006
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2007"
            >
              2007
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2008"
            >
              2008
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2009"
            >
              2009
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2010"
            >
              2010
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2011"
            >
              2011
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2012"
            >
              2012
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2013"
            >
              2013
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2014"
            >
              2014
            </option>
            <option
              class="sc-gqHWMn gwEFlI"
              value="2015"
            >
              2015
            </option>
          </select>
          <p
            class="sc-SYvCl csQMFq"
          />
        </div>
      </div>
      <input
        class="sc-hnHaRn eKyvOD"
        data-testid="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <p
        class="sc-SYvCl csQMFq"
      />
      <div
        class="sc-QUirw jnedLj"
      >
        <button
          class="sc-erFrCC iFaZZO"
          data-testid="submit-button"
        >
          Sign up
        </button>
      </div>
    </form>
  </section>
</div>
`;
